epochStart/metachain/pendingMiniBlocks.go

```diff
@@ -1,257 +1,126 @@
-		return nil, err
-	}
-
-	var metaHeader block.MetaBlock
-	err = p.marshalizer.Unmarshal(&metaHeader, buff)
-	if err != nil {
-		return nil, err
 	}
 
-	return &metaHeader, nil
+	return crossShardMiniBlocks
 }
 
 // AddProcessedHeader will add all miniblocks headers in a map
-func (p *pendingMiniBlockHeaders) AddProcessedHeader(handler data.HeaderHandler) error {
-	if check.IfNil(handler) {
+func (p *pendingMiniBlocksHeaders) AddProcessedHeader(headerHandler data.HeaderHandler) error {
+	if check.IfNil(headerHandler) {
 		return epochStart.ErrNilHeaderHandler
 	}
 
-	metaHdr, ok := handler.(*block.MetaBlock)
+	metaBlock, ok := headerHandler.(*block.MetaBlock)
 	if !ok {
 		return epochStart.ErrWrongTypeAssertion
 	}
 
-	crossShard := p.getAllCrossShardMiniBlocks(metaHdr)
+	crossShardMiniBlocks := p.getAllCrossShardMiniBlocks(metaBlock)
 
-	var err error
 	p.mutPending.Lock()
-	defer func() {
-		p.mutPending.Unlock()
-		if err != nil {
-			_ = p.RevertHeader(handler)
-		}
-	}()
-
-	for key, mbHeader := range crossShard {
-		if _, ok = p.mapMiniBlockHeaders[key]; !ok {
-			p.mapMiniBlockHeaders[key] = mbHeader
+	for mbHash, mbHeader := range crossShardMiniBlocks {
+		if _, ok = p.mapMiniBlockHeaders[mbHash]; !ok {
+			p.mapMiniBlockHeaders[mbHash] = mbHeader
 			p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID]++
 			continue
 		}
 
-		delete(p.mapMiniBlockHeaders, key)
-		p.decrementNumMiniBlocks(mbHeader.ReceiverShardID)
-
-		var buff []byte
-		buff, err = p.marshalizer.Marshal(&mbHeader)
-		if err != nil {
-			return err
+		delete(p.mapMiniBlockHeaders, mbHash)
+		if p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID] > 0 {
+			p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID]--
 		}
+	}
 
-		err = p.storage.Put(mbHeader.Hash, buff)
-		if err != nil {
-			return err
-		}
+	for shardID, numPendingMiniBlocks := range p.mapShardNumMiniBlocks {
+		log.Trace("pending miniblocks", "shard", shardID, "num", numPendingMiniBlocks)
 	}
 
+	p.mutPending.Unlock()
+
 	return nil
 }
 
 // RevertHeader will remove all miniblocks headers that are in metablock from pending
-func (p *pendingMiniBlockHeaders) RevertHeader(handler data.HeaderHandler) error {
-	if check.IfNil(handler) {
+func (p *pendingMiniBlocksHeaders) RevertHeader(headerHandler data.HeaderHandler) error {
+	if check.IfNil(headerHandler) {
 		return epochStart.ErrNilHeaderHandler
 	}
 
-	metaHdr, ok := handler.(*block.MetaBlock)
+	metaBlock, ok := headerHandler.(*block.MetaBlock)
 	if !ok {
 		return epochStart.ErrWrongTypeAssertion
 	}
 
-	crossShard := p.getAllCrossShardMiniBlocks(metaHdr)
+	crossShardMiniBlocks := p.getAllCrossShardMiniBlocks(metaBlock)
 
 	p.mutPending.Lock()
-	for mbHash, mbHeader := range crossShard {
+	for mbHash, mbHeader := range crossShardMiniBlocks {
 		if _, ok = p.mapMiniBlockHeaders[mbHash]; ok {
 			delete(p.mapMiniBlockHeaders, mbHash)
-			p.decrementNumMiniBlocks(mbHeader.ReceiverShardID)
+			if p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID] > 0 {
```

Comment:
can we log an error if we deleted but the num miniblocks was 0?

Reply:
Done